### PR TITLE
Adds OpenAPI spec, adds missing 400s in spec, fixes bookingig in spec, updates http status codes in the code

### PR DESCRIPTION
**Pull request by joelrosario**
This PR does the following:

- Adds OpenAPIValidator middleware to the project to return 400 responses when any request does not match the specification
- Adds an OpenAPI 3.0.3 specification which is used by the above validator middleware
- Updated the application to return 400 / 404 status codes consistent with RESTful API design where appropriate, and updated the specification to reflect this
- PUT /booking/{id} used to throw an exception if a spec-valid request of the form {"task": {.. booking properties.. } } was sent to it, and this is fixed in this PR
- Updated bookingid from string to integer in the specification (specification made it a string but the application returned a number)
- Added mandatory keys into the required list in their respective objects

### CODE REVIEW

File: routes/index.js [hunk 1]
```diff
@@ -120,7 +120,7 @@ router.put('/booking/:id', function(req, res, next) {
     updatedBooking = req.body;
     if(req.headers['content-type'] === 'text/xml') updatedBooking = updatedBooking.booking;
 
-    validator.scrubAndValidate(updatedBooking, function(payload, msg){
+    validator.scrubAndValidate(updatedBooking.task, function(payload, msg){
```
Comment (emilybache) [inline routes/index.js:123]: I believe that in this case it's the schema which is in error and the code which is correct.

File: oas.yaml [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+            UPDATE_BOOKING:
+              value: Basic YWRtaW46cGFzc3dvcmQxMjM=
+      requestBody:
+        description: booking object
+        content:
+          '*/*':
+            schema:
+              type: object
+              required:
+                - task
```
Comment (emilybache) [inline oas.yaml:262]: It is my belief that Mark has deliberately inserted this inconsistency into the spec in order to provide a more interesting exercise for testers. I think the schema for 'PUT' should be the same as for 'CREATE', ie this 'task' is an unnecessary complication

File: swagger.json [hunk 3]
```diff
@@ -272,7 +313,8 @@
                         "description": "The authorization header. There is a secret value that only application administrators know.",
                         "schema": {
                           "type": "string"
-                        }
+                        },
+                        "required": true
```
Comment (emilybache) [inline swagger.json:317]: I don't believe this header is actually required. You can also log in using a cookie.

File: swagger.json [hunk 4]
```diff
@@ -180,7 +204,8 @@
...
```
Comment (emilybache) [inline swagger.json:208]: I don't believe this header is actually required. You can also log in using a cookie.

**Comment by emilybache** [comment]
Thankyou for taking the time to look at this code! Some of your changes are useful and I would like to include. Others I am less keen to have. Let me go through them.

- You've suggested some useful improvements to swagger.json. I hadn't thought it worth including the 400 and 404 errors in it but it's probably better to have them
- The authorization header I don't believe is required - if you have set the login cookie it won't be checked.
- I'm mostly wanting to keep all the bugs in index.js. You've fixed a few of them but there are plenty more. I'd like to keep it as a better exercise for testers, ie with more bugs.
- Because I want to keep the bugs, I think I might not include the code to check against the schema. Perhaps I will include it commented out. In a real system I'd want to have it, but in a real system I'd also want to fix the bugs :-)

**Comment by emilybache** [comment]
I've been looking at this more closely and I wondered about the oas.yaml file you provided in this pull request. I was under the impression you got it from Mark's original repo but I couldn't find it there. Can you point me at a location on the internet where I can get this file or should I assume you designed it yourself? If you designed it yourself why did you add the 'task' attribute to the booking PUT endpoint?

**Comment by joelrosario** [comment]
1. Yes indeed, we got it from Mark's repo. Here is the [link](https://github.com/texttest/restful-booker/blob/with_texttests/swagger.json) to that file. However you'll notice that this is in the outdated Swagger 2.0 format. We converted it into OpenAPI 3.0 format and saved it as oas.yaml.
2. As you might recall, the specification was out of sync with the application, which is why we had to update it to get the contract tests passing.

**Pull request merged**